tokio/src/time/interval.rs

```diff
@@ -399,6 +400,10 @@ pub struct Interval {
     /// The strategy `Interval` should use when a tick is missed.
     missed_tick_behavior: MissedTickBehavior,
 
+    /// The previous period is used to restore the period after
+    /// using `reset_immediately`, `reset_after` or `reset_at`.
+    prev_period: Option<Duration>,
```

Comment:
I don't really understand why you added this field. What situation breaks without it?

Reply:
I assume that this is so that, when you use `reset_after(delay)`, the Interval acts as though it has an period of `delay`, instead of its real period; the real period then needs to be stored somewhere, so that it can be restored afterwards.

Perhaps this is so that, if a task wants to get an estimate for when the next tick would happen, and it tries to derive it from the Interval's period [1], that it would get a value that's not affected by any `reset_after`'s. (For example, normally you get a tick every 100ms, but then you do a `reset_after(Duration::from_secs(5))`, and then you get confused when you "miss" 50 ticks.)

Personally, I think that events such as `reset_after` are exceptional, in that they don't reflect anything new about the Interval itself, so they don't need to mutate it except where technically necessary (i.e. `self.delay` needs to be `reset`). If you wanted to change when the ticks happen on a permanent basis, you would replace this Interval with a new one.

[1] The fact that there doesn't seem to be an easy way to get the projected Instant of the next tick, meaning you need to rely on weird hacks like these, is itself a bit of a problem, but that's out of scope for this issue.

Reply:
Okay, but as far as I can tell, this isn't actually the case. Calling `reset_after` does not modify the `period` field.

Reply:
`Sleep::reset_inner` calls `TimerEntry::reset(deadline, true)` which mutates the internal period.

Although, now that I see more carefully there is a `Sleep::reset_without_reregister` which doesn't update the period. I could try using this.

Reply:
Not sure I understand. The `TimerEntry` can't even access the `period` field, since that field is outside it.

I would recommend that you come up with a test that shows why it is broken without `prev_period`.

Reply:
It turns out it was "broken" only on `reset_at` when the deadline was in the past.
The first tick after reset would happen 0s after, and the second would happen at "period - difference on reset"

```rust
use tokio::time::{self, Instant};

use std::time::Duration;

#[tokio::main]
async fn main() {
    let start_time = Instant::now();
    let mut interval = time::interval(Duration::from_millis(100));

    interval.tick().await;
    println!("Expected: `0s` elapsed, actual: `{:?}`", start_time.elapsed());

    time::sleep(Duration::from_millis(50)).await;
    println!("Expected: `50s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.reset_immediately_no_prev_period();

    interval.tick().await;
    println!("Expected: `0s + 50s = 50s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.tick().await;
    println!("Expected: `100s + 50s = 150s` elapsed, actual: `{:?}`", start_time.elapsed());
}
```

Output:
```sh
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/examples/interval-at-negative-no-prev-period`
Expected: `0s` elapsed, actual: `1.137366ms`
Expected: `50s` elapsed, actual: `52.44776ms`
Expected: `0s + 50s = 50s` elapsed, actual: `52.468425ms`
Expected: `100s + 50s = 150s` elapsed, actual: `103.02993ms`
```

Reply:
So I guess there's no need for prev period, I'll just have to fix the `reset_at`.